### ISSUE TITLE
508-defect-1 ❗ Launchblocker [COGNITION] Sitewide: Multi-step forms MUST announce page headings to screen readers when users press Back or Continue

### ISSUE THREAD

**Issue by 1Copenut** [department-of-veterans-affairs/va.gov-team#12323]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

**Feedback framework**

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Description

Today I discovered a bug that iOS Safari + VoiceOver was not announcing the H2 of multi-step forms when users pressed the Continue button. This lead to an additional discovery that screen readers are not announcing the H2 reliably when users press Back or Continue. This is a significant hinderance to screen reader users and I believe could be fixed by moving the tabIndex and focusing on the heading instead of the containing div. Screenshot and code sample attached below.

## Point of Contact

**VFS Point of Contact:** _Trevor_

## Acceptance Criteria

- [ ] Screen readers announce the top H2 on every route change of multi-step forms
- [ ] Focus is set properly on the H2 on every route change
- [ ] H2 should not have the yellow focus halo. (This may change in the future, but for now the guidance is no halo on non-native focused elements.)

## Possible Fixes (optional)

This issue should be resolved by moving the `tabIndex="-1"` to the H2 and calling `focus()` on that element.

```diff
<div
  aria-valuemax="5"
  aria-valuemin="1"
  aria-valuenow="1"
  aria-valuetext="Step 1 of 5: Applicant Information"
  class="nav-header nav-header-schemaform"
- tabindex="-1"
+ >
    <h2
      class="vads-u-font-size--h4"
+     tabindex="-1"
+   >
      <span class="form-process-step current">1</span>
      <span class="form-process-total">of 5</span>
    Applicant Information
  </h2>
</div>
```

## WCAG or Vendor Guidance (optional)

* [Info and Relationships: Understanding SC 1.3.1](https://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-programmatic.html)

## Screenshots or Trace Logs

![Screen Shot 2020-08-10 at 4 04 05 PM](https://user-images.githubusercontent.com/934879/89832374-d4940f00-db24-11ea-8373-97409528a9f9.png)

### Prescribed change:
![image](https://user-images.githubusercontent.com/12564977/91046882-32c1f700-e5e7-11ea-93fd-cc36c0cad386.png)

**Comment by rianfowler** [comment]
Focusing the `h2` and moving the tabIndex doesn't resolve this issue. This old [post](https://stackoverflow.com/questions/9250847/advancing-voiceover-to-a-specified-element-in-mobile-safari) suggests forcing VO to focus an `h2` doesn't work on mobile safari. 

Is there something else we can try like moving the aria attributes?

**Comment by 1Copenut** [comment]
@rianfowler & @meganhkelley 
After doing a lot of testing with iOS and desktop screen readers, I've been able to isolate the problem in a Codepen: https://codepen.io/tpierce_402/pen/PoNZbzQ

> The good news is we can now say we understand exactly what’s causing the problem. If the container DIV is focused, it cannot have siblings to the heading element. If the heading is focused, it cannot have child elements like spans. I captured all of that using dynamic and static tabIndex declarations, being careful to only change one variable per test.

LMK when we've got a working update in Staging and I'll test in IE11

**Comment by 1Copenut** [comment]
@emilywaggoner Rian and I are working to solve this one. The issue appears to be caused by having spans inside the heading for visual distinction. Looks like we might need to remove the spans and styling, but wanted to get your input before greenlighting that decision.

**Comment by emilywaggoner** [comment]
Thanks for the heads up @1Copenut! Is the screenshot below an accurate representation of what it would look like with out the spans?

<img width="673" alt="Screen Shot 2020-08-18 at 1 43 50 PM" src="https://user-images.githubusercontent.com/1379629/90547163-117d8880-e159-11ea-88c1-efa9826d947c.png">

**Comment by 1Copenut** [comment]
@emilywaggoner Yes, that's what our H2 would look like without any child spans. 

I opened a bug ticket with Apple using their Feedback system here: https://feedbackassistant.apple.com/feedback/8452615 . Not sure if that will get resolved or just be an item in the backlog, but figured it couldn't hurt to approach this as a forms lib item, and a bug in iOS Safari VoiceOver.

**Comment by emilywaggoner** [comment]
Got it - thanks! I think the styling is fine, but I think it merits another look at the copy. Without the specialized styling, it reads a little funny as a heading, so I imagine we might want to make a slight change to something like, "Step 1 of 6: Veteran Information" 

I would love to consult @peggygannon on what actual copy improvements we might want to make. Peggy, would love your input -- let me know if you need additional context for the changes being made here!

@1Copenut just to be totally clear, this will be a global change on all browsers and operating systems, not just the Apple ones?

**Comment by 1Copenut** [comment]
@emilywaggoner 
> this will be a global change on all browsers and operating systems, not just the Apple ones?

Correct, this would be an update to the HTML and CSS, so all browsers would pick up the single heading style. I'd test it thoroughly in the major (and minor) browser pairings in staging before giving it a 👍

**Comment by peggygannon** [comment]
thanks, @emilywaggoner and @1Copenut So to make sure I'm understanding correctly, the update would be to remove the specialized styling (blue ) in these chapter headers? 

If so, I agree with Emily's suggestion above, adding colon btw. the step and the chapter title. 
Step 1 of 6: Veteran information

Would this update apply for all form chapter heads?

**Comment by emilywaggoner** [comment]
I think it would apply to all form chapter heads, but I'll let @1Copenut confirm.

Also wanted to loop in @rtwell for visibility since this is a minor, but global design change.

**Comment by 1Copenut** [comment]
Yes, this change would apply to all form headers under the progress bar. I think the actual term for those is form chapter heads?... This is a big small context and visual change; I appreciate the guidance and awareness you're all bringing to the issue. Making a change like this only works when it doesn't degrade the experience for some while improving it for others.

**Comment by rtwell** [comment]
+1 to @peggygannon's suggestion—describing the chapters/sections is most helpful. tyvm @emilywaggoner for looping me in!

**Comment by rtwell** [comment]
donezo 🎩

**Comment by meganhkelley** [comment]
Hey y'all! I'm looking for a recap on next steps for this one. Is this global design system changed approved, and FE Tools should implement? Or is this waiting to go to the design system council for approval? 

@rtwell @emilywaggoner @1Copenut @peggygannon

**Comment by KevinMHoffmanUSDS** [comment]
This global design system change is approved, and FE should implement.

**Comment by 1Copenut** [comment]
@bkjohnson && @meganhkelley 
I've reviewed this on Staging, using several screen reader and browser combos. Results are as follows:

* Win10 + IE11 + JAWS 2020 - Reads heading when previous page has an update. No degradation from existing.
* Win10 + Chrome + JAWS 2020 - Reads heading when previous page has an update. No degradation from existing.
* Win10 + Edge Chromium + JAWS 2020 - Reads heading when previous page has an update. No degradation from existing.
* Win10 + Firefox + NVDA - Reads heading correctly
* Win10 + Chrome + NVDA - Reads heading correctly
* MacOS + Safari + VoiceOver - Reads heading correctly (tested on localhost feature branch)
* iOS + Safari + VoiceOver - Reads heading correctly (tested on localhost feature branch)

I'm ready to close this issue. Thank you both and @cvalarida for all the work and patience to find a good solution that keeps the behavior consistent for most, and improves it for mobile users.
